modules/videoio/src/cap_libcamera.cpp

```diff
@@ -0,0 +1,1001 @@
+bool LibcameraCapture::grabFrame()
+{
+    if (!camerastarted)
+    {
+        if (!startVideo())
+        {
+            CV_LOG_ERROR(NULL, "VIDEOIO(Libcamera): Failed to start camera");
+            return false;
+        }
+    }
+    
+    // Try to get a message from libcamera (non-blocking check)
+    try 
+    {
+        // Use Wait() with short timeout to check for available messages
+        LibcameraCapture::Msg msg = Wait();
+
+        if (msg.type == LibcameraCapture::MsgType::RequestComplete)
+        {
+            CompletedRequestPtr payload = msg.getCompletedRequest();
+            {
+                std::lock_guard<std::mutex> lock(request_mutex_);
+                current_request_ = payload;
+            }
+            return true;
+        }
+        else if (msg.type == LibcameraCapture::MsgType::Quit)
+        {
+            CV_LOG_INFO(NULL, "VIDEOIO(Libcamera): Quit message received");
+            return false;
+        }
+    }
+    catch (const std::exception& e)
+    {
+        // No message available or error occurred
+        return false;
+    }
+    
+    return false;
+}
+
+/**
+ * @brief Retrieve the frame data from the current completed request.
+ *
+ * This function extracts frame data directly from libcamera's memory-mapped buffers
+ * without additional copying. It uses the completed request obtained in grabFrame().
+ *
+ * @param int Unused parameter.
+ * @param dst An OpenCV `OutputArray` where the retrieved frame will be stored.
+ *            The frame is stored in RGB format (8-bit, 3 channels, CV_8UC3).
+ *
+ * @return `true` if a frame is successfully retrieved and copied to `dst`.
+ *         `false` if no frame is ready or an error occurred.
+ */
+bool LibcameraCapture::retrieveFrame(int, OutputArray dst)
+{
+    CompletedRequestPtr request;
+    {
+        std::lock_guard<std::mutex> lock(request_mutex_);
+        if (!current_request_)
+        {
+            return false;
+        }
+        request = current_request_;
+        current_request_ = nullptr; // Clear after use
+    }
+    
+    // Get the viewfinder stream
+    libcamera::Stream *stream = ViewfinderStream(&vw, &vh, &vstr);
+    if (!stream)
+    {
+        CV_LOG_ERROR(NULL, "VIDEOIO(Libcamera): Error getting viewfinder stream");
+        return false;
+    }
+    
+    // Get memory mapped buffer directly from libcamera
+    std::vector<libcamera::Span<uint8_t>> mem = Mmap(request->buffers[stream]);
+    if (mem.empty())
+    {
+        CV_LOG_ERROR(NULL, "VIDEOIO(Libcamera): Error getting memory mapped buffer");
+        return false;
+    }
+    
+    // Create OpenCV Mat directly from libcamera buffer
+    Mat frame(vh, vw, CV_8UC3);
+    uint8_t *src_ptr = mem[0].data();
+    uint line_size = vw * 3;
+    
+    // Copy line by line to handle stride correctly
+    for (unsigned int i = 0; i < vh; i++, src_ptr += vstr)
+    {
+        memcpy(frame.ptr(i), src_ptr, line_size);
+    }
+    
+    frame.copyTo(dst);
+    return true;
```

Comment:
You have 2 copies here. cv::Mat supports stride (step) you can construct cv::Mar from `mem[0].data()` and call copyTo for it.